pkg/asset/machines/openstack/machines.go

```diff
@@ -37,24 +37,34 @@ func Machines(clusterID string, config *types.InstallConfig, pool *types.Machine
 	if poolPlatform := pool.Platform.Name(); poolPlatform != openstack.Name {
 		return nil, fmt.Errorf("non-OpenStack machine-pool: %q", poolPlatform)
 	}
-	platform := config.Platform.OpenStack
-
-	az := ""
-	provider, err := generateProvider(clusterID, platform, pool.Platform.OpenStack, osImage, az, role, userDataSecret)
-	if err != nil {
-		return nil, err
-	}
 
-	if role == "master" {
-		provider.ServerGroupName = clusterID + "-master"
-	}
+	mpool := pool.Platform.OpenStack
+	platform := config.Platform.OpenStack
 
 	total := int64(1)
 	if pool.Replicas != nil {
 		total = *pool.Replicas
 	}
 	machines := make([]machineapi.Machine, 0, total)
 	for idx := int64(0); idx < total; idx++ {
+		zone := mpool.Zones[int(idx)%len(mpool.Zones)]
+		provider, err := generateProvider(
```

Comment:
So we generate a spec per machine? Shouldn't we instead generate a spec per availability zone? Or even better, move the `checkNetworkExtensionAvailability()` func making call to the OpenStack API out of `generateProvider()` since calling is multiple times is not expected to give different results.

Reply:
I'll see what I can do here. It would be nice to minimize the duplication of api calls here as much as possible

Reply:
Ok, should be settled in the latest commit